osutil/disks/disks_linux.go

```diff
@@ -379,51 +400,14 @@ func (d *disk) Schema() string {
 	return d.schema
 }
 
-// diskFromMountPointImpl returns a Disk for the underlying mount source of the
-// specified mount point. For mount points which have sources that are not
-// partitions, and thus are a part of a disk, the returned Disk refers to the
-// volume/disk of the mount point itself.
-func diskFromMountPointImpl(mountpoint string, opts *Options) (*disk, error) {
-	// first get the mount entry for the mountpoint
-	mounts, err := osutil.LoadMountInfo()
-	if err != nil {
-		return nil, err
-	}
-	var d *disk
-	var partMountPointSource string
-	// loop over the mount entries in reverse order to prevent shadowing of a
-	// particular mount on top of another one
-	for i := len(mounts) - 1; i >= 0; i-- {
-		if mounts[i].MountDir == mountpoint {
-			d = &disk{
-				major: mounts[i].DevMajor,
-				minor: mounts[i].DevMinor,
-			}
-			partMountPointSource = mounts[i].MountSource
-			break
-		}
-	}
-	if d == nil {
-		return nil, fmt.Errorf("cannot find mountpoint %q", mountpoint)
-	}
-
-	// now we have the partition for this mountpoint, we need to tie that back
-	// to a disk with a major minor, so query udev with the mount source path
-	// of the mountpoint for properties
-	props, err := udevPropertiesForName(partMountPointSource)
-	if err != nil && props == nil {
-		// only fail here if props is nil, if it's available we validate it
-		// below
-		return nil, fmt.Errorf("cannot find disk for partition %s: %v", partMountPointSource, err)
-	}
-
+func diskFromPartUdevPropsAndOpts(props map[string]string, opts *Options) (*disk, error) {
 	if opts != nil && opts.IsDecryptedDevice {
```

Comment:
minor: since you are refactoring anyway, it would be nice to have a separate `udevPropsForDecryptedDevice()` function which does the steps currently done inside this `if` block.

Reply:
Can be done in a followup.

Reply:
Also note we have https://github.com/snapcore/snapd/pull/11028 which will also significantly re-write this function, so it would be nice to not introduce more conflicts until things have settled here a bit more. I think with that PR though this function is a bit more clean, but we could still split out this logic to a helper in that PR